Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,627 @@
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+class TextAreaBox::ContentSizedTextEdit : public QTextEdit {
+  public:
+
+    explicit ContentSizedTextEdit(const QString& text,
+        QWidget* parent = nullptr)
+        : QTextEdit(parent) {
+      setLineWrapMode(QTextEdit::WidgetWidth);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+      document()->setDocumentMargin(0);
+      setFrameShape(QFrame::NoFrame);
+      setAcceptRichText(false);
+      connect(
+        this, &QTextEdit::textChanged, this, [=] { on_text_changed(); });
+      setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+      setText(text);
+    }
+
+    QSize sizeHint() const override {
+      auto margins = contentsMargins();
+      auto desired_size =
+        QSize(m_longest_line_width, document()->size().toSize().height());
+      auto size = desired_size + QSize(margins.left() + margins.right(),
+        margins.top() + margins.bottom());
+      if(!isReadOnly()) {
+        size.rwidth() += cursorWidth();
+      }
+      return size;
+    }
+
+    QSize minimumSizeHint() const override {
+      return QSize();
+    }
+
+  private:
+    int m_longest_line_width;
+
+    void on_text_changed() {
+      m_longest_line_width = get_longest_line_width();
+      updateGeometry();
+    }
+
+    int get_longest_line_width() const {
+      auto longest = 0;
+      for(auto i = 0; i < document()->blockCount(); ++i) {
+        auto block = document()->findBlockByNumber(i);
+        if(block.isValid()) {
+          longest =
+            std::max(longest, fontMetrics().horizontalAdvance(block.text()));
+        }
+      }
+      return longest;
+    }
+};
+
+class TextAreaBox::ElidedLabel : public QWidget {
+  public:
+
+    explicit ElidedLabel(QWidget *parent = nullptr)
+      : QWidget(parent) {}
+
+    void set_text(const QString &text) {
+      m_text = text;
+      update();
+    }
+
+    void set_text_color(const QColor& color) {
+      m_text_color = color;
+      update();
+    }
+
+    void set_alignment(Qt::Alignment alignment) {
+      m_alignment = alignment;
+      update();
+    }
+
+    void set_line_height(int line_height) {
+      m_line_height = line_height;
+      update();
+    }
+
+    QSize sizeHint() const override {
+      return {width(), height()};
+    }
+
+  protected:
+    void paintEvent(QPaintEvent *event) override {
+      auto painter = QPainter(this);
+      painter.setPen(m_text_color);
+      painter.setFont(font());
+      auto fontMetrics = painter.fontMetrics();
+      auto y = m_line_height - fontMetrics.height();
+      auto textLayout = QTextLayout(m_text, painter.font());
+      auto opt = textLayout.textOption();
+      opt.setAlignment(m_alignment);
+      textLayout.setTextOption(opt);
+      textLayout.beginLayout();
+      auto line = textLayout.createLine();
+      while(line.isValid()) {
```

Comment:
This appears like a for loop:

    for(auto line = foo(); line.isValid(); line = foo()) { ... }